pkg/controller/nativedaemonset/nativedaemonset_controller.go

```diff
@@ -0,0 +1,267 @@
+type ReconcileNativeDaemonSet struct {
+	// client interface
+	client.Client
+	eventRecorder record.EventRecorder
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to DaemonSet with specific annotations
+	updateHandler := func(e event.UpdateEvent) bool {
+		oldObject := e.ObjectOld.(*appsv1.DaemonSet)
+		newObject := e.ObjectNew.(*appsv1.DaemonSet)
+		if len(oldObject.Annotations) != len(newObject.Annotations) || !reflect.DeepEqual(oldObject.Annotations, newObject.Annotations) {
+			klog.V(3).Infof("Observed updated Annotation for native DaemonSet: %s/%s", newObject.Namespace, newObject.Name)
+			return true
+		}
+		return false
+	}
+
+	// Watch for changes to DaemonSet
+	return c.Watch(source.Kind(mgr.GetCache(), &appsv1.DaemonSet{}), &handler.EnqueueRequestForObject{}, predicate.Funcs{UpdateFunc: updateHandler})
+}
+
+// Reconcile reads that state of the cluster for a DaemonSet object and makes changes based on the annotations
+// to implement progressive delivery by deleting pods according to batch requirements.
+func (r *ReconcileNativeDaemonSet) Reconcile(_ context.Context, request reconcile.Request) (reconcile.Result, error) {
+	daemon := new(appsv1.DaemonSet)
+	err := r.Get(context.TODO(), request.NamespacedName, daemon)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Check for partition annotation
+	partitionStr, hasPartition := daemon.Annotations[util.DaemonSetPartitionAnnotation]
+	if !hasPartition {
+		// No partition annotation, nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	// Get all pods owned by this DaemonSet using ListOwnedPods utility function
+	pods, err := util.ListOwnedPods(r.Client, daemon)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Calculate desired updated replicas based on partition, which equals to total pods number - partition
+	desiredUpdatedReplicas, err := daemonsetutil.CalculateDesiredUpdatedReplicas(partitionStr, len(pods))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	klog.Infof("Processing DaemonSet %s/%s with desired updated replicas: %d", daemon.Namespace, daemon.Name, desiredUpdatedReplicas)
+
+	// Process the pods according to partition requirements
+	result, err := r.processBatch(daemon, pods, desiredUpdatedReplicas)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return result, nil
+}
+
+// analyzePods analyzes the current pod state and determines how many pods need to be deleted
+// Returns: podsToDelete slice, needToDelete count, batchCompleted bool, error
```

Comment:
batchCompleted seems redundant, if needToDelete is zero, batchCompleted is always true